### ISSUE TITLE
Odd mouse detection and sound problems in 3D Logic

### ISSUE THREAD

**Issue by WumboSpasm** [ruffle-rs/ruffle#1149]
https://chat.kongregate.com/gamez/0000/0179/live/3d_logic.swf

You're supposed to be able to click anywhere within a tile to activate it, but when played through Ruffle, the areas where you can activate a tile changes depending on the position of the tile (as far as I can tell). Most of the time you can only click specific sections, ranging from the entire inner part to only a couple pixels. Dragging is fine but it sometimes activates a different tile than the one you're hovering over.

As for the sound, the intro music is supposed to fade in and then fade out once you click start. Instead, it starts at full volume and keeps looping after the start button is clicked. In addition, the level completion jingle lacks the audio mixing present in Adobe's players (it's supposed to play forward and backward while the cube is fading out and in respectively but it only plays forward).

(and one more teeny tiny thing: the outlines of the marker tiles' inner squares are missing the bottom side)

**Comment by kmeisthax** [comment]
Initial triage: This game appears to have significant performance issues due to it's dynamic rendering. It also hit-tests on it's dynamically drawn shapes, which should be fine, but is obviously entirely broken on Ruffle. `useHandCursor` is also not being respected, which is actually useful as it lets me see the underlying issue in real-time.

I figured it would be something to do with clip depth and having lots of hit-tested shapes partially intersecting each other's bounding boxes. So I tried disabling the bounding-box check in `hitTest`, which just resulted in clicking a polygon coloring half the cube. I dived deeper into the drawing API that powers hit tests, and there's a lot of comments like this:

```
// TODO: Handle cases where fill is not closed.
// Probably should have an explicit `flush` method that handles this.
```

![image](https://user-images.githubusercontent.com/108736/94328791-65af3000-ff83-11ea-97c5-427485844ec1.png)

At least some of these shapes are definitely not closed, just by looking at them. However, inspecting `com.Polygon.draw`, it issues `beginFill`, `moveTo`, `lineTo`, and `endFill` commands as one would expect. These shapes *should* be closed. I dropped the game into Flash Player for a look, and it does appear to close the strokes correctly.

So this is likely to be missing functionality in Ruffle's internal drawing API. The game will issue points A, B, C, and D; and then expect Flash to draw the line from D back to A, which we apparently don't do. Our hit tests work using winding numbers, where we count how many parts of the line are to our left or right in order to determine if we're inside it. If we're missing a line, then it creates an entire horizontal deadzone where the shapetest will spuriously fail or succeed when it shouldn't.

The offending code is in `core/src/drawing.rs`. It seems to have several TODOs relating to closing shapes, and it seems like this game has caught us forgetting to implement that.

(And now I just noticed that you already pointed out the missing line. It's most certainly not "teeny tiny", as that is a visual manifestation of the underlying bug.)

The audio issues are due to us not having implementations for several `Sound` methods, such as `setVolume`.
